### PR TITLE
Package ometrics.0.1.1

### PR DESCRIPTION
**Pull request by vch9**
### `ometrics.0.1.1`
OCaml analysis in a merge request changes

---
* Homepage: https://gitlab.com/nomadic-labs/ometrics
* Source repo: git+https://gitlab.com/nomadic-labs/ometrics.git
* Bug tracker: https://gitlab.com/nomadic-labs/ometrics/-/issues

---
:camel: Pull-request generated by opam-publish v2.1.0

### CODE REVIEW

**Comment by vch9** [comment]
I added several dependencies restrictions, but I'm not sure whether its the right way to do.

File: packages/ometrics/ometrics.0.1.1/opam [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+bug-reports: "https://gitlab.com/nomadic-labs/ometrics/-/issues"
+synopsis: "OCaml analysis in a merge request changes"
+
+depends: [
+  "ocaml" {>= "4.12" & < "4.13"}
+  "dune" {>= "2.9.1"}
+  "yojson" {>= "1.7.0"}
+  "menhirSdk" {>= "20211012" & < "20211013"}
+  "menhirLib" {>= "20211012" & < "20211013"}
+  "menhir" {>= "20211012" & < "20211013"}
```
Comment (mseri) [inline packages/ometrics/ometrics.0.1.1/opam:15]: Are you sure you need so strict bounds here?

Comment (mseri) [reply packages/ometrics/ometrics.0.1.1/opam:15]: Oh, I see. If you are vendoring merlin, you should instead add a conflict:
```
conflict: [
  "menhir"
]
```

In general (as far as I know) there can be issues/complex conflicts when you have a dependency both installed and vendored.

If you think only a specific set of versions have issues, you can do
```
conflict: [
  "menhir" {!= "20211012"}
]
```

**Comment by kit-ty-kate** [comment]
> > ```
> > #=== ERROR while compiling ometrics.0.1.1 =====================================#
> > # context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.12.1 | pinned(https://gitlab.com/nomadic-labs/ometrics/-/archive/0.1.1/ometrics-0.1.1.tar.gz)
> > # path                 ~/.opam/4.12/.opam-switch/build/ometrics.0.1.1
> > # command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ometrics -j 31
> > # exit-code            1
> > # env-file             ~/.opam/log/ometrics-4830-0a5907.env
> > # output-file          ~/.opam/log/ometrics-4830-0a5907.out
> > ### output ###
> > # File "lib/dune", line 3, characters 35-50:
> > # 3 |  (libraries unix str merlin.kernel merlin.analysis)
> > #                                        ^^^^^^^^^^^^^^^
> > # Error: Library "merlin.analysis" not found.
> > # Hint: try:
> > #   dune external-lib-deps --missing --no-config --root . --ignore-promoted-rules --default-target @install --always-show-command-line --promote-install-files --release --only-packages ometrics -p ometrics --profile release -j 31 @install
> > ```
> 
> It's part of the vendored lib, isn't opam installing git submodules (`git submodule update --init`)? I already used the executable in tests with a simple `opam pin add ometrics git+https://gitlab.com/nomadic-labs/ometrics.git#main`. At first I also had `ometrics` as a public library but I ran into [ocaml/dune#5301](https://github.com/ocaml/dune/issues/5301).

The archive contains the submodule, but `dune` automatically disables all vendored packages when `-p` is used. I believe in your case you might have to use `--release` only instead.

File: packages/ometrics/ometrics.0.1.1/opam (hunk 1 above)
Comment (vch9) [reply packages/ometrics/ometrics.0.1.1/opam:15]: I first (locally) tested with
```
conflicts: [
  "menhir"
]
```
but I got the same error about versions. It worked fine with your second suggestion.

**Comment by vch9** [comment]
Thank you both @mseri @kit-ty-kate , I applied both of your suggestions and tested the installation locally, went fine :). Waiting for the CI.

**Comment by vch9** [comment]
> > > ```
> > > #=== ERROR while compiling ometrics.0.1.1 =====================================#
> > > # context              2.0.10 | linux/x86_64 | ocaml-base-compiler.4.12.1 | pinned(https://gitlab.com/nomadic-labs/ometrics/-/archive/0.1.1/ometrics-0.1.1.tar.gz)
> > > # path                 ~/.opam/4.12/.opam-switch/build/ometrics.0.1.1
> > > # command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p ometrics -j 31
> > > # exit-code            1
> > > # env-file             ~/.opam/log/ometrics-4830-0a5907.env
> > > # output-file          ~/.opam/log/ometrics-4830-0a5907.out
> > > ### output ###
> > > # File "lib/dune", line 3, characters 35-50:
> > > # 3 |  (libraries unix str merlin.kernel merlin.analysis)
> > > #                                        ^^^^^^^^^^^^^^^
> > > # Error: Library "merlin.analysis" not found.
> > > # Hint: try:
> > > #   dune external-lib-deps --missing --no-config --root . --ignore-promoted-rules --default-target @install --always-show-command-line --promote-install-files --release --only-packages ometrics -p ometrics --profile release -j 31 @install
> > > ```
> > 
> > 
> > It's part of the vendored lib, isn't opam installing git submodules (`git submodule update --init`)? I already used the executable in tests with a simple `opam pin add ometrics git+https://gitlab.com/nomadic-labs/ometrics.git#main`. At first I also had `ometrics` as a public library but I ran into [ocaml/dune#5301](https://github.com/ocaml/dune/issues/5301).
> 
> The archive contains the submodule, but `dune` automatically disables all vendored packages when `-p` is used. I believe in your case you might have to use `--release` only instead.

According to the CI replacing `-p name` by `--release` does not seem to work.

**Comment by kit-ty-kate** [comment]
> The archive contains the submodule

Seems like I was wrong about that. Automatically generated images from gitlab don’t seem to contain their submodules, you’d have to upload your own archive with all the submodules

**Comment by vch9** [comment]
> > The archive contains the submodule
> 
> Seems like I was wrong about that. Automatically generated images from gitlab don’t seem to contain their submodules, you’d have to upload your own archive with all the submodules

Thank you for investigating. Do you think it would be a good idea to mirror the project on github to facilitate publications in the future?

**Comment by vch9** [comment]
Regarding submodules, @clarus pointed me this issue https://github.com/dear-github/dear-github/issues/214. This can be a frequent issue if you encounter others developers using sub modules.

**Comment by vch9** [comment]
Seems to work!

**Comment by kit-ty-kate** [comment]
Thanks!

**Comment by kit-ty-kate** [comment]
Btw for next time you might want to use dune-release which generates the archive locally (with all the submodules) and uploads it

**Comment by vch9** [comment]
> Btw for next time you might want to use dune-release which generates the archive locally (with all the submodules) and uploads it

I've been ignoring dune-release for too long! Thank you for the help!

**Pull request merged**